### ISSUE TITLE
orphaned files on spool

### ISSUE THREAD

**Issue by brianjmurrell** [cyrusimap/cyrus-imapd#2919]
I believe I have a pile of orphaned files in my imapd spool.  If I count the files in a given folder:

```bash
# ls /var/spool/imap/b/user/brian/[redacted]/ | wc -l
125003

```

Yet multiple imap client say there are only 24862 messages in this folder.

Is there a more definitive way to get cyrus imapd's understanding of this folder than using an imap client?

**Comment by dilyanpalauzov** [comment]
Maybe the files are not expunged yet, see cyr_expire and from imapd.conf the option expunge_mode.

**Comment by brianjmurrell** [comment]
```bash
# grep expunge /etc/imapd.conf
```

But the messages in this folder that are not being shown to an imap client are years old.  I'd hope that they'd have been expunged by now.

**Comment by dilyanpalauzov** [comment]
You have to call cyr_expire.  Please read the manual.

**Comment by brianjmurrell** [comment]
Why do I have to call `cyr_expire` for these messages when I routinely delete messages in imap folders and they are removed immediately from the cyrus imap spool without me having to call `cyr_expire` for them?  My understanding is that this immediate removal, without waiting for cyr_expire to be run is the default behaviour when not specified otherwise for 2.4.17, is it not?

**Comment by dilyanpalauzov** [comment]
Yes, the default in 2.4 for expunge_mode is immediate/default.  Later the default was changed to delayed.

I do not know why you have this effect.

**Comment by brianjmurrell** [comment]
So what would be the next step in determining if these files really are orphans and can be safely deleted?  And if so, which all are orphans and which are not?

**Comment by dilyanpalauzov** [comment]
At your place as next step I would upgrade to a newer cyrus Imap version, then run cyr_expire and see if the problem remains.

**Comment by dilyanpalauzov** [comment]
You can also try with `reconstruct`.

**Comment by brianjmurrell** [comment]
`reconstruct` is interesting.  I had forgotten about that.  Does the imap server need to be stopped before I run it?

Is it typically pretty quiet, even if it does need to do some rebuilding?  I.e.:

```bash
# /usr/lib/cyrus-imapd/reconstruct -n -f user.brian.[redacted]
user.brian.[redacted]
```

**Comment by dilyanpalauzov** [comment]
I am not sure.  Maybe you do not have to stop the server, as there are sufficent mutex mechanisms available.

You can read the available options of reconstruct in the manual and attach to a running reconstruct with sphinx to check which file is currently read.  I will likely not give you more information than is documented, but somebody could give a different answer.  You can also write to the users mailing list.

**Comment by dilyanpalauzov** [comment]
Per https://www.cyrusimap.org/imap/reference/manpages/systemcommands/cyradm.html reconstruct can be started over cyradm.  This means that the server does not have to be shut down.

**Comment by brianjmurrell** [comment]
So, further to this, if I choose a file in my spool which I think is an orphan, such as `/var/spool/imap/b/user/brian/1446888` and then run `/usr/lib/cyrus-imapd/mbexamine user.brian` and search for `1446888` and don't find it in that output, it's an orphan, isn't it?

**Comment by brianjmurrell** [comment]
Further investigation seems to reveal the above cross-referencing is valid.  And if so, I certainly do have orphaned files in my mail spool.

Previous comments indicate that reconstruct is supposed to be able to resolve this.  Is this true.  Would I want it's `-O` option to remove these orphan message files?  `-n` is an interesting option, but since it doesn't actually report anything at all about what `reconstruct` is (or would be without `-n`) doing it of pretty little value.

**Comment by rsto** [comment]
This issue hasn't seen any updates since 5 years. If it still shows up to be an issue on a supported Cyrus version (3.8, 3.10), please reopen this issue.
